### PR TITLE
[GlobalISel] Fix the infinite loop issue in `commute_int_constant_to_rhs`

### PR DESCRIPTION
**Pull request by darkbuck**
- When both operands are constant, the matcher runs into an infinite
  loop as the commutation should be applied only when LHS is a constant
  and RHS is not.

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
+# RUN: llc -mtriple aarch64 -global-isel -run-pass=aarch64-prelegalizer-combiner -verify-machineinstrs %s -o - \
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir:2]:
```suggestion
# RUN: llc -mtriple aarch64 -run-pass=aarch64-prelegalizer-combiner %s -o - \
```

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# RUN:     --aarch64prelegalizercombiner-disable-rule=constant_fold_binop | FileCheck %s
+
+# `constant_fold_binop` is disabled to trigger the infinite loop in `commute_int_constant_to_rhs`.
+
+---
+name:            add
+tracksRegLiveness: true
+body:             |
+  bb.1:
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir:11]:
```suggestion
  bb.0:
```

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    ; CHECK-NEXT: %add:_(s32) = G_ADD %c0, %c1
+    ; CHECK-NEXT: $s0 = COPY %add(s32)
+    ; CHECK-NEXT: RET_ReallyLR
+    %c0:_(s32) = G_CONSTANT i32 1
+    %c1:_(s32) = G_CONSTANT i32 2
+    %add:_(s32) = G_ADD %c0, %c1
+    $s0 = COPY %add
+    RET_ReallyLR
+
+...
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir:28]: Maybe add a vector version?

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir:1]: Can you merge this test with test/CodeGen/AArch64/GlobalISel/combine-const-fold-barrier-rhs.mir

Comment (darkbuck) [reply llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir:1]: > Can you merge this test with test/CodeGen/AArch64/GlobalISel/combine-const-fold-barrier-rhs.mir

This test needs `--aarch64prelegalizercombiner-disable-rule=constant_fold_binop` to trigger issue. That makes the merge complicated. Shall we leave as it is.

**Review by arsenm — APPROVED** [review]

**Comment by RKSimon** [comment]
@darkbuck This is failing on some buildbots - please can you take a look? https://lab.llvm.org/buildbot/#/builders/124/builds/10375

**Comment by annamthomas** [comment]
@darkbuck  We are seeing the same failing tests on our downstream builds because of this test failure: https://lab.llvm.org/buildbot/#/builders/124/builds/10375
Given one of the buildbots failed (and I'm not sure what the fix is here), can we please revert to green? https://llvm.org/docs/DeveloperPolicy.html#patch-reversion-policy

**Comment by gulfemsavrun** [comment]
This test also failed in our Clang toolchain builders.
```
FAIL: LLVM :: CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir (2372 of 53803)
******************** TEST 'LLVM :: CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir' FAILED ********************
Exit Code: 2

Command Output (stderr):
--
RUN: at line 2: /b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/bin/llc -mtriple aarch64 -run-pass=aarch64-prelegalizer-combiner /b/s/w/ir/x/w/llvm-llvm-project/llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir -o -      --aarch64prelegalizercombiner-disable-rule=constant_fold_binop | /b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/bin/FileCheck /b/s/w/ir/x/w/llvm-llvm-project/llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir
+ /b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/bin/FileCheck /b/s/w/ir/x/w/llvm-llvm-project/llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir
+ /b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/bin/llc -mtriple aarch64 -run-pass=aarch64-prelegalizer-combiner /b/s/w/ir/x/w/llvm-llvm-project/llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir -o - --aarch64prelegalizercombiner-disable-rule=constant_fold_binop
LLVM ERROR: Invalid rule identifier
FileCheck error: '<stdin>' is empty.
FileCheck command line:  /b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/bin/FileCheck /b/s/w/ir/x/w/llvm-llvm-project/llvm/test/CodeGen/AArch64/GlobalISel/combine-commute-int-const-lhs.mir

--
```
https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8751607530180046481/+/u/clang/test/stdout

**Comment by annamthomas** [comment]
Okay, I'll try reverting the change given these buildbot failures. @darkbuck  pls feel free to reland with fix(es).

**Pull request merged**